File: 1657/determine-if-two-strings-are-close.py
# 兩字串close：交換2字母位置 or 將字母的值交換
# abcde -> aecdb 把 b e 交換位置
# aacabb -> bbcbaa 把 a b 的值交換
# 因為字的長度很長 10^5 不能用暴力法，也不知要怎麼暴力
# 可能與「統計字母出現次數」有關。只要把次數排序，即可
class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        if len(word1)!=len(word2): return False
        # 長度不同，就不能可能成功

        H1 = defaultdict(int)
        H2 = defaultdict(int) # 統計字母出現次數
        for c in word1: H1[c] += 1
        for c in word2: H2[c] += 1
        # 前面統計，後面查看集合 & 次數排序
        key1 = {k for k in H1.keys()}
        key2 = {k for k in H2.keys()}
        if key1 != key2: return False
        # 統計的字母集合不同，也是失敗

        count1 = [v for v in H1.values()]
        count2 = [v for v in H2.values()]

        count1.sort()
        count2.sort()
        for i in range(len(count1)): # 小到大的個數不同，也失敗
            if count1[i]!=count2[i]: return False
        return True # 經過嚴格考驗，總算成功了
# case 131/153: "uau", "ssx" # 字母不同，也不行
# case 132/153: "abbzzca", "babzzcz" 
# aabbczz vs. bbzzzc

File: 1657/determine-if-two-strings-are-close.cpp
// LeetCode 1657. Determine if Two Strings Are Close
// 它的操作，是任意交換位置，還可以任意字母對調翻轉。所以統計好數量，並看字母set相同，就成功
class Solution {
public:
    bool closeStrings(string word1, string word2) {
        if(word1.length() != word2.length()) return false; // 長度要相同
        vector<int> counter1(26,0); // 26個字母，預設值 0
        vector<int> counter2(26,0); // 26個字母，預設值 0
        for(char c : word1) counter1[c-'a']++; // 統計字母出現次數
        for(char c : word2) counter2[c-'a']++; // 統計字母出現次數

        for(int i=0; i<26; i++){ // 字母「一邊有、一邊沒有」，就失敗
            if(counter1[i]!=0 && counter2[i]==0) return false;
            if(counter1[i]==0 && counter2[i]!=0) return false;
        }
        sort(counter1.begin(), counter1.end()); // 照頻率排序
        sort(counter2.begin(), counter2.end()); // 照頻率排序
        for(int i=0; i<26; i++) { // 若數量不同，就失敗
            if(counter1[i] != counter2[i]) return false;
        }
        return true;
    }
};
